netsync/manager.go

```diff
@@ -840,23 +987,128 @@ func (sm *SyncManager) handleBlockMsg(bmsg *blockMsg) {
 		}
 	}
 
-	// If we are not in headers first mode, it's a good time to periodically
-	// flush the blockchain cache because we don't expect new blocks immediately.
-	// After that, there is nothing more to do.
-	if !sm.headersFirstMode {
-		if err := sm.chain.FlushUtxoCache(blockchain.FlushPeriodic); err != nil {
-			log.Errorf("Error while flushing the blockchain cache: %v", err)
-		}
+	// It's a good time to periodically flush the blockchain cache because
+	// we don't expect new blocks immediately.  After that, there is
+	// nothing more to do.
+	if err := sm.chain.FlushUtxoCache(blockchain.FlushPeriodic); err != nil {
+		log.Errorf("Error while flushing the blockchain cache: %v", err)
+	}
+}
+
+// handleBlockMsgHeadersFirst handles block messages from all peers when the
+// sync manager is in headers first mode.  For blocks received out of order, it
+// first keeps them in memory and sends them to be processed when the next block
+// from the tip is available.
+func (sm *SyncManager) handleBlockMsgHeadersFirst(bmsg *blockMsg) {
+	peer := bmsg.peer
+	_, exists := sm.peerStates[peer]
+	if !exists {
+		log.Warnf("Received block message from unknown peer %s", peer)
 		return
 	}
+	blockHash := bmsg.block.Hash()
+
+	// Add the block to the queue.
+	sm.queuedBlocks[*blockHash] = bmsg
+
+	// Log the progress as we have received the block.
+	sm.lastProgressTime = time.Now()
+
+	firstNodeEl := sm.headerList.Front()
+	if firstNodeEl == nil {
+		log.Errorf("missing first node in the headerlist on block %v",
+			bmsg.block.Hash())
+		return
+	}
+
+	// Look for blocks that we can process next.
+	processBlocks := make([]*blockMsg, 0, len(sm.queuedBlocks)+1)
+	var next *list.Element
+	for e := firstNodeEl; e != nil; e = next {
+		next = e.Next()
+		node, ok := e.Value.(*headerNode)
+		if !ok {
+			log.Warn("Header list node type is not a headerNode")
+			continue
+		}
+
+		b, found := sm.queuedBlocks[*node.hash]
+		if !found {
+			// Break when we're missing the next block in
+			// sequence.
+			break
+		}
+
+		// We have a block we can process so remove from the queue and
+		// add it to the processBlocks slice.
+		processBlocks = append(processBlocks, b)
+		delete(sm.queuedBlocks, *node.hash)
+
+		// Remove the block from the headerList.
+		//
+		// NOTE We leave in the checkpointed hash so that we can connect
+		// headers on the next get headers request.
+		if !node.hash.IsEqual(sm.nextCheckpoint.Hash) {
+			sm.headerList.Remove(e)
+		}
+	}
+
+	// If we have any blocks to process, process them now.
+	isCheckpointBlock := false
+	for _, processBlock := range processBlocks {
+		// Process the block to include validation, best chain selection, orphan
+		// handling, etc.
+		_, _, err := sm.chain.ProcessBlock(processBlock.block, blockchain.BFFastAdd)
+		if err != nil {
```

Comment:
I think there exists a DoS vector that the peer can send us the same orphan block twice and shut down our node.

Reply:
I don't think they can since we ignore blocks that we never requested during headers-first but I'll add a test to make sure it does not fail.